server/etcdserver/api/v3rpc/watch.go

```diff
@@ -432,11 +445,15 @@ func (sws *serverWatchStream) sendLoop() {
 				Canceled:        canceled,
 			}
 
-			if _, okID := ids[wresp.WatchID]; !okID {
-				// buffer if id not yet announced
-				wrs := append(pending[wresp.WatchID], wr)
-				pending[wresp.WatchID] = wrs
-				continue
+			// Progress notifications can have WatchID -1
+			// if they announce on behalf of multiple watchers
+			if !progressNotify || wresp.WatchID != clientv3.InvalidWatchID {
```

Comment:
Why do we need to check both `!progressNotify` and `wresp.WatchID != clientv3.InvalidWatchID`? 
Based on the comment above a would conclude that `progressNotify => (wresp.WatchID == clientv3.InvalidWatchID)`.

Reply:
Progress notifications *only* have WatchID -1 *if* they are announced on behalf of multiple workers. The counter-example is timed progress notifications, which are specific to a worker, and could therefore be buffered here.

The other edge case here would be a non-notification (i.e. actual payload) with a watch ID of -1, which I am reasonably (?) sure never happen. After all, this would lead to them getting "buffered" forever. So I guess I agree, I'll fix that.